cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/spec/PipelineSpecGenerator.java

```diff
@@ -642,6 +642,21 @@ protected ValidatedPipeline validateConfig(ETLConfig config) {
       }
     }
 
+    // make sure sources are not in the middle of the pipeline after any non-condition/action stage
+    for (Map.Entry<String, ETLStage> entry : stages.entrySet()) {
```

Comment:
instead of another loop through the dag, can we do this check in the previous loop (line 596)?